packages/teuchos/core/src/Teuchos_TestForException.cpp

```diff
@@ -109,3 +110,8 @@ bool Teuchos::TestForException_getEnableStacktrace()
 {
   return loc_enableStackTrace();
 }
+
+void Teuchos::TestForTermination_terminate(const std::string &msg) {
+  std::cerr << msg;
+  std::terminate();
```

Comment:
When MPI is enabled, should this not be MPI_Abort() instead of std::terminate()?

Reply:
I actually spent a good amount of time considering `MPI_Abort`, and I don't think its the right choice here. The most important reason is covered in [this blog post](https://akrzemi1.wordpress.com/2011/10/05/using-stdterminate/) while explaining why `std::terminate()` cannot call `std::exit()`:

"This poses a risk that in case std::terminate is called as a consequence of throwing exception from global destructor (say, the very last global destructor), calling std::exit would launch the process of calling all destructors again… Therefore, C++ Standard defines the situation of calling std::exit after the clean-up of globals has started, an undefined behavior."

If `TEUCHOS_TEST_FOR_TERMINATION` were to call anything that calls `exit()` (and `MPI_Abort()` does indeed call `exit()`), then it would invoke this undefined behavior if it gets triggered from a global destructor, and it exists primarily for error handling within destructors.

Reply:
There is no mention at all of MPI in [that blog post](https://akrzemi1.wordpress.com/2011/10/05/using-stdterminate/).  I would be concerned that having just one process calling `std::terminate()` would result in the other MPI processes hanging.  To avoid that, I would rather just call MPI_Abort() and guarantee that the whole MPI program will come down and just burn the whole thing down.

I guess I could write some multi-process tests that calls this macro and then see what happens.  I just worry that some of those tests might hang with some C++ and MPI  implementation pairs.

What do you think?

Reply:
I can't find any hits on Google for the expected or proper interaction of `std::terminate()` with MPI programs.  But from looking at [this item](http://comp.parallel.mpi.narkive.com/1XiQoFZH/calling-exit-or-mpi-abort-in-an-mpi-program), we should expect modern MPI implementations and runtime environments to kill the other MPI processes of one MPI process is terminated (either by exit() or abort()).

So hopefully my tests will show that calling std::terminate() on one process of a multi-process system will bring down all of the processes.

But then should we not at least give the MPI proc-ID in the std::cerr output line in debugging?  We can get that from Teuchos::GlobalMPISession.  I can add that to the automated tests.

I will push an updated commit to your branch.

Reply:
I looked at the source code of MPICH and it calls `exit()`. Every MPI implementation I've ever dealt with would terminate all other ranks if one of them stops in any way, including `abort()` (which is what my library usually does when something goes wrong) or even segfault, or throwing an exception like `std::bad_alloc` when out of memory. Most of them also give an indication of which rank (or ranks) ended badly, even if Teuchos doesn't do it explicitly. I don't mind trying to print the global rank though.

Reply:
Also this branch is on my fork, I've invited you as a collaborator so we don't have to do tedious things like another pull request into the branch.

Reply:
@ibaned, turns out that you don't need push access to people's GitHub clones anymore.  It turns out that a while back that GitHub implemented a feature that when you create a PR, that you by default given permission for the developers of that target GitHub repo to push to your branch.  Next time you create a PR, look for the check-box where you are giving this permission.

Reply:
@ibaned ok that makes sense

Reply:
@bartlettroscoe  oh, I didn't know that included commits. I checked that box for this PR, so we should be good.